tests/journal/balance-assertions.test

```diff
@@ -136,3 +136,140 @@ hledger -f - stats
 # >>> /Transactions/
 # >>>2
 # >>>=0
+
```

Comment:
Oops, submitted before I was finished.. still learning github's review flow. 

Overall I like this a lot. I think it is not easy code for future hledger hackers to understand, eg at handleObject I gave up (for now). But I'm not sure much can be done about it; the feature is inherently complex. It seems to me you've done a good job encapsulating everything in a principled and compact way, and perhaps we will find small improvements in future making it more transparent. And your tests give confidence. I'll do some additional testing of my own next. Great work @johannesgerer.